src/EditorFeatures/Core/Diagnostics/DiagnosticsSquiggleTaggerProvider.cs

```diff
@@ -47,11 +48,16 @@ protected sealed override bool SupportsDiagnosticMode(DiagnosticMode mode)
 
         protected sealed override bool IncludeDiagnostic(DiagnosticData diagnostic)
         {
-            var isUnnecessary = diagnostic.Severity == DiagnosticSeverity.Hidden && diagnostic.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary);
+            if (!string.IsNullOrWhiteSpace(diagnostic.Message))
+            {
+                if (diagnostic.Severity is DiagnosticSeverity.Warning or DiagnosticSeverity.Error)
+                    return true;
+
+                if (diagnostic.Severity == DiagnosticSeverity.Hidden && diagnostic.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary))
+                    return true;
+            }
 
-            return
-                (diagnostic.Severity == DiagnosticSeverity.Warning || diagnostic.Severity == DiagnosticSeverity.Error || isUnnecessary) &&
-                !string.IsNullOrWhiteSpace(diagnostic.Message);
+            return false;
```

Comment:
this logic should likely be moved to a shared location with LSP pull diags as well.